### PR TITLE
Add `subsets(collection,Val{k})`

### PR DESCRIPTION
**Pull request by ettersi**
Implements #12. 

Performance shoot-out:

```
# Reference implementation for listing all pairs
julia> function collect_pairs(x)
           p = Vector{NTuple{2,eltype(x)}}(binomial(length(x),2))
           idx = 1
           for i = 1:length(x)
               for j = i+1:length(x)
                   p[idx] = (x[i],x[j])
                   idx += 1
               end
           end
           return p
       end

# Small collections
julia> c = collect(1:10)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}));
  190.040 ns (1 allocation: 816 bytes)
  3.490 μs (140 allocations: 6.95 KiB)
  1.642 μs (22 allocations: 1.44 KiB)

# Medium-size collections
julia> c = collect(1:100)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}));
  12.735 μs (2 allocations: 77.45 KiB)
  296.917 μs (14857 allocations: 735.06 KiB)
  4.759 μs (204 allocations: 83.73 KiB)

# Large collections
julia> c = collect(1:1000)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}));
  1.319 ms (2 allocations: 7.62 MiB)
  74.309 ms (1498509 allocations: 72.41 MiB)
  32.392 μs (2004 allocations: 7.68 MiB)

```

Conclusions: 

- Could be a bit faster for small collections, very performant otherwise
- There are more allocations than there should be. Curiously, the number of allocations scales linearly in the size of the input collection, not quadratically. Couldn't really figure out where they come from.

### CODE REVIEW

**Comment by ettersi** [comment]
I forgot to update the REQUIRE to include the dependency on StaticArrays. How can I add a commit to this PR? (Sorry for being a noob, this is my first ever PR)

File: src/IterTools.jl [hunk 1]
```diff
@@ -708,6 +708,50 @@ end
 done(it::Binomial, state::BinomialIterState) = state.done
 
 
+# Iterate over all subsets of a collection with a given *statically* known size
+
+struct StaticSizeBinomial{K,Container}
+    xs::Container
+end
+
+iteratorsize(::Type{<:StaticSizeBinomial}) = HasLength()
```
Comment (iamed2) [inline src/IterTools.jl]: Only if the container has length; see the `Subsets` type. Also define `iteratoreltype` as this will only have eltype if `Container` has eltype.

File: src/IterTools.jl [hunk 2]
```diff
@@ -708,6 +708,50 @@ end
...
+
+eltype(::Type{StaticSizeBinomial{K,C}}) where {K,C} = NTuple{K,eltype(C)}
+length(it::StaticSizeBinomial{K,C}) where {K,C} = binomial(length(it.xs),K)
+
+subsets(xs,::Type{Val{K}}) where {K} = StaticSizeBinomial{K,typeof(xs)}(xs)
```
Comment (iamed2) [inline src/IterTools.jl]: Would it be worth replacing `subsets(xs, k::Integer)` with this implementation as well? It wouldn't be as fast as this but might it be as fast or faster than the `Binomial` implementation?

File: src/IterTools.jl [hunk 3]
```diff
@@ -708,6 +708,50 @@ end
...
+
+using StaticArrays
+function start(it::StaticSizeBinomial{K,C}) where {K,C}
```
Comment (iamed2) [inline src/IterTools.jl]: Leave the parameters off the type when they're not used; `C` is not used here and in other places, and neither `K` nor `C` is used in `done`.

File: src/IterTools.jl [hunk 4]
```diff
@@ -708,6 +708,50 @@ end
...
+    n = length(it.xs)
+    return MVector((K <= n ? 0 : 1, ntuple(i->i,Val{K})...))
```
Comment (iamed2) [inline src/IterTools.jl]: Use `ntuple(identity, Val{K}())`

File: src/IterTools.jl (hunk 2 above)
Comment (iamed2) [inline src/IterTools.jl]: This should be `subsets(xs, ::Val{K})` instead (you'll have to adjust the tests).

File: src/IterTools.jl [hunk 5]
```diff
@@ -708,6 +708,50 @@ end
...
+end
+
+function next(it::StaticSizeBinomial{K,C}, idx) where {K,C}
+    xs = it.xs
+    sidx = ((_,idx...)->idx)(idx.data...) # Type-stable version of idx.data[2:end]
+    x = map(i->xs[i], sidx)
```
[the remark this answers is not in the document]
Comment (iamed2) [reply src/IterTools.jl]: (note that the above two changes might change the `eltype`, which should be reflected elsewhere)

File: src/IterTools.jl (hunk 1 above)
Comment (ettersi) [reply src/IterTools.jl]: I was following the implementation of `Binomial`, which has the same issues. The point with `iteratoreltype` is fair enough and should be changed for all three cases of `subsets`. All three implementations require `length(xs)`, though, so all `subsets` iterators have a length if they're defined.

File: src/IterTools.jl (hunk 2 above)
Comment (ettersi) [reply src/IterTools.jl]: Maybe, although I am not sure whether it is worth the effort. Imagine you want to list all subsets of size `n-1` of a set of size `n`, for `n = O(10_000)`. That means you would have tuples of length `O(n)` in your code, which seems odd. And it would be a breaking change, so a bit painful to do.

Comment (ettersi) [reply src/IterTools.jl]: The `subsets(xs, ::Val{K})` syntax is from 0.7 onwards, and mere mortals like me are still on 0.6. This will be a trivial change once 0.7 gets officially released.

File: src/IterTools.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (ettersi) [reply src/IterTools.jl]: It doesn't seem to work:
```
julia> (1:5)[(2,4,3)]
ERROR: ArgumentError: invalid index: (2, 4, 3)
Stacktrace:
 [1] getindex(::UnitRange{Int64}, ::Tuple{Int64,Int64,Int64}) at ./abstractarray.jl:883

julia> collect(1:5)[(2,4,3)]
ERROR: ArgumentError: invalid index: (2, 4, 3)
Stacktrace:
 [1] getindex(::Array{Int64,1}, ::Tuple{Int64,Int64,Int64}) at ./abstractarray.jl:883

```

File: src/IterTools.jl [hunk 6]
```diff
@@ -708,6 +708,50 @@ end
...
+
+    begin # i = findlast(i->idx[i] != length(xs)-K+i-1, 2:K+1)+1
```
[the remark this answers is not in the document]
Comment (ettersi) [reply src/IterTools.jl]: No. I am using the code blocks to indicate that these bits of code are the type-stable versions of the commented snippets.

File: src/IterTools.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (ettersi) [reply src/IterTools.jl]: Ah, it does work if we index with a `StaticArray`. Changed it to `x = xs[shift(idx)].data`. I think we should anyway return a tuple and not a `StaticArray` here.

File: src/IterTools.jl (hunk 3 above)
Comment (ettersi) [reply src/IterTools.jl]: Updated. Still getting used to the `<:Type` syntax.

File: src/IterTools.jl (hunk 4 above)
Comment (ettersi) [reply src/IterTools.jl]: Updated.

File: src/IterTools.jl [hunk 7]
```diff
@@ -708,6 +708,50 @@ end
...
+
+using StaticArrays
+function start(it::StaticSizeBinomial{K,C}) where {K,C}
+    n = length(it.xs)
+    return MVector((K <= n ? 0 : 1, ntuple(i->i,Val{K})...))
+end
+
+function next(it::StaticSizeBinomial{K,C}, idx) where {K,C}
+    xs = it.xs
+    sidx = ((_,idx...)->idx)(idx.data...) # Type-stable version of idx.data[2:end]
```
[the remark this answers is not in the document]
Comment (ettersi) [reply src/IterTools.jl]: Updated (see below).

File: src/IterTools.jl (hunk 1 above)
Comment (ettersi) [reply src/IterTools.jl]: It seems several iterators in this package do not define `iteratoreltype` correctly. So it might be worth opening up another PR to fix this for all.

Comment (iamed2) [reply src/IterTools.jl]: Binomial doesn't have the same issues, as it only accepts a `Vector` which always has an eltype and length.

File: src/IterTools.jl (hunk 2 above)
Comment (iamed2) [reply src/IterTools.jl]: Right, that makes sense. Perhaps include a comment about the tradeoffs?

Comment (iamed2) [reply src/IterTools.jl]: `subsets(xs, ::Val{K})` works just fine on 0.6. You just need to pass `Val{K}()`.

**Comment by ettersi** [comment]
All the above requests should be addressed now. Unfortunately, I found a bug in my code which meant that I effectively only iterated over all subsets of size `K-1`. Bug is fixed now and additional tests have been added to check for this, but the performance dropped by a factor `n`, obviously:

```
julia> c = collect(1:10)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  184.855 ns (1 allocation: 816 bytes)
  3.505 μs (140 allocations: 6.95 KiB)
  1.198 μs (49 allocations: 2.28 KiB)

julia> c = collect(1:100)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  12.973 μs (2 allocations: 77.45 KiB)
  295.724 μs (14857 allocations: 735.06 KiB)
  83.270 μs (4955 allocations: 232.22 KiB)

julia> c = collect(1:1000)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  1.311 ms (2 allocations: 7.62 MiB)
  85.998 ms (1498509 allocations: 72.41 MiB)
  9.641 ms (499505 allocations: 22.87 MiB)
``` 

I further found the culprit for the extra allocations: `MVector` is heap-allocated, hence returning it in a tuple in `next` causes the tuple to be heap-allocated as well. This is hard to avoid, however, and I doubt it is very important for performance. 

In conclusion, this implementation of `subsets(c,Val{k}())` is significantly slower than hand-writing the loop, which is a bit of a bummer. I don't see how this can be avoided, however.

**Comment by ettersi** [comment]
Found an alternative implementation which avoids `MVector`s and hence gets rid of the extra allocations and is slightly faster. 

```
julia> c = collect(1:10)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  200.645 ns (1 allocation: 816 bytes)
  3.895 μs (140 allocations: 6.95 KiB)
  965.000 ns (3 allocations: 864 bytes)

julia> c = collect(1:100)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  13.004 μs (2 allocations: 77.45 KiB)
  334.811 μs (14857 allocations: 735.06 KiB)
  56.056 μs (4 allocations: 77.50 KiB)

julia> c = collect(1:1000)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  1.309 ms (2 allocations: 7.62 MiB)
  80.856 ms (1498509 allocations: 72.41 MiB)
  5.597 ms (4 allocations: 7.62 MiB)
```

I believe there is no inherent reason why the current implementation should be significantly slower than the explicit loops, for we mostly do the same operations in the inner-most loop: evaluating a single `xs[i]`, assembling the return tuple (or can the compiler eliminate this?), incrementing the index, and checking whether we are done with the innermost loop (technically this happens twice (once in `next` and once in `done`), but I again hope the compiler can eliminate this). So I assume one could still speed things up a bit by giving some hints to the compiler, but I leave this to others as I have very little experience with that.

**Comment by ettersi** [comment]
Found an alternative implementation which avoids `MVector`s and hence gets rid of the extra allocations. After throwing in a few `@inline`s this is as fast as we could hope for: 
```
julia> function collect_pairs(x)
                  p = Vector{NTuple{2,eltype(x)}}(binomial(length(x),2))
                  idx = 1
                  for i = 1:length(x)
                      for j = i+1:length(x)
                          p[idx] = (x[i],x[j])
                          idx += 1
                      end
                  end
                  return p
              end
collect_pairs (generic function with 1 method)

julia> c = collect(1:10)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  192.050 ns (1 allocation: 816 bytes)
  3.527 μs (140 allocations: 6.95 KiB)
  589.261 ns (3 allocations: 864 bytes)

julia> c = collect(1:100)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  12.546 μs (2 allocations: 77.45 KiB)
  301.494 μs (14857 allocations: 735.06 KiB)
  15.313 μs (4 allocations: 77.50 KiB)

julia> c = collect(1:1000)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  1.305 ms (2 allocations: 7.62 MiB)
  84.349 ms (1498509 allocations: 72.41 MiB)
  1.504 ms (4 allocations: 7.62 MiB)
```

File: src/IterTools.jl [hunk 8]
```diff
@@ -722,34 +722,47 @@ length(it::StaticSizeBinomial{K,<:Any}) where {K} = binomial(length(it.xs),K)
 
 subsets(xs,::Val{K}) where {K} = StaticSizeBinomial{K,typeof(xs)}(xs)
 
-using StaticArrays
-function start(it::StaticSizeBinomial{K,<:Any}) where {K}
-    n = length(it.xs)
-    return MVector((K <= n ? 0 : 1, ntuple(identity,Val{K}())...))
-end
+@generated minus1(::Val{A}) where {A} = :(Val{$(A-1)}())
```
Comment (iamed2) [inline src/IterTools.jl]: Jameson has said "no" to this in the past. However, you can use `first` and `Base.tail` for inferred head and tail on tuples.

**Comment by iamed2** [comment]
Did you benchmark this on 0.7 as well? Inference has changed in ways that might affect this sort of code (it did in AxisArrays, for example).

**Comment by ettersi** [comment]
I'm trying to benchmark on 0.7, but `JSON` seems to be broken which prevents me from using BenchmarkTools. Is there a workaround?

File: src/IterTools.jl (hunk 8 above)
Comment (ettersi) [reply src/IterTools.jl]: Replaced `pop()` with
```
@inline pop(t::NTuple) = reverse(Base.tail(reverse(t))), t[end]
```
Don't really see the problem with the old code, though.

**Comment by iamed2** [comment]
I think you could checkout this branch: https://github.com/JuliaIO/JSON.jl/pull/226

Or wait for it to be merged.

**Comment by ettersi** [comment]
I added documentation for the new method and benchmarked on 0.7. The performance gets worse indeed. 

```
julia> c = collect(1:10)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  194.970 ns (1 allocation: 816 bytes)
  2.209 μs (49 allocations: 4.81 KiB)
  5.268 μs (2 allocations: 832 bytes)

julia> c = collect(1:100)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  10.179 μs (2 allocations: 77.45 KiB)
  207.946 μs (4955 allocations: 502.98 KiB)
  20.339 μs (3 allocations: 77.47 KiB)

julia> c = collect(1:1000)
       @btime collect_pairs(c);
       @btime collect(subsets(c,2));
       @btime collect(subsets(c,Val{2}()));
  1.019 ms (2 allocations: 7.62 MiB)
  51.525 ms (499506 allocations: 49.54 MiB)
  1.401 ms (3 allocations: 7.62 MiB)
```

**Comment by ettersi** [comment]
Sorry, I realised I never came round to fix `iteratorelsize`. I believe everything should be ready now, and there's nothing left for me to do. Please let me know if I'm wrong.

**Comment by iamed2** [comment]
Could you rebase to resolve the conflicts? I'll revisit this today.

**Comment by ettersi** [comment]
I believe I resolved the conflicts, but as it is the first time of me doing this I might have well screwed up some things.

**Comment by iamed2** [comment]
It turns out the `@inline`s had no effect on performance, so I removed them. I also cleaned up the code. I will merge today unless you have concerns about the `@inline`s.

Thanks for your perseverance!

**Comment by ettersi** [comment]
Nope, I forgot when and why I put the `@inline`, so I am happy to remove them if you checked that this doesn't impact performance. 

Thank you for the mentoring, it's been very instructive for me!

**Comment by iamed2** [comment]
Yup, ran all the same benchmarks.

You're welcome!!
